### DISCUSSION TITLE
Given N numbers in an array, your task is to keep at most the top K numbers with respect to their frequency.

### DISCUSSION THREAD

**Discussion by RoboScriptSmith** [getlost01/gfg-potd#42] [Solution Ideas]
class Solution {
public:
    vector<vector<int>> kTop(vector<int>& arr, int N, int K) {
        vector<vector<int>> result;

        // vector of size K+1 to store elements
        vector<int> top(K + 1);

        // array to keep track of frequency
        unordered_map<int, int> freq;

        // iterate till the end of stream
        for (int m = 0; m < N; m++) {
            // increase the frequency
            freq[arr[m]]++;

            // store that element in top vector
            top[K] = arr[m];

            // search in top vector for the same element
            auto it = find(top.begin(), top.end() - 1, arr[m]);

            // iterate from the position of the element to zero
            for (int i = distance(top.begin(), it) - 1; i >= 0; --i) {
                // compare the frequency and swap if a higher frequency element is stored next to it
                if (freq[top[i]] < freq[top[i + 1]])
                    swap(top[i], top[i + 1]);

                // if frequency is the same, compare the elements and swap if the next element is higher
                else if ((freq[top[i]] == freq[top[i + 1]]) && (top[i] > top[i + 1]))
                    swap(top[i], top[i + 1]);
                else
                    break;
            }

            // create a temporary vector to store the current top K elements
            vector<int> currentTop;
            for (int i = 0; i < K && top[i] != 0; ++i)
                currentTop.push_back(top[i]);

            // add the current top K elements to the result vector
            result.push_back(currentTop);
        }

        return result;
    }
};

**Comment by Addy0710** [comment]
class Solution {
  public:
    vector<vector<int>> kTop(vector<int>& arr, int N, int K) {
        vector<int> top(K + 1,0);
        vector<vector<int>> ans;
        unordered_map<int, int> mp;
        for (int i = 0; i < N; i++) {
            mp[arr[i]]++;
            top[K] = arr[i];
            auto it = find(top.begin(), top.end() - 1, arr[i]);
            int index = it - top.begin() - 1;
            for (int j = index ; j >= 0; --j) {
                if (mp[top[j]] < mp[top[j + 1]])
                    swap(top[j], top[j + 1]);
                else if ((mp[top[j]] == mp[top[j + 1]]) && (top[j] > top[j+1]))
                    swap(top[j], top[j + 1]);
                else
                    break;
            }
            vector<int>temp;
            for (int i = 0; i < K && top[i] != 0; ++i)
                temp.push_back(top[i]);
            ans.push_back(temp);
        }
        return ans;
    }
};


//this works
